python/hail/expr/types.py

```diff
@@ -172,14 +172,26 @@ def _add_jtype(self, jtype):
     def _has_jtype(self):
         return self._cached_jtype is not None
 
-    @abc.abstractmethod
-    def _typecheck(self, annotation):
-        """
-        Raise an exception if the given annotation is not the appropriate type.
+    def typecheck(self, value):
+        """Check that `value` matches a type.
+
+        Parameters
+        ----------
+        value
+            Value to check.
 
-        :param annotation: value to check
+        Raises
+        ------
+        :obj:`TypeError`
         """
-        return
+        def check(t, obj):
+            t._typecheck(obj)
+            return True
+        self._traverse(value, check)
+
+    @abc.abstractmethod
+    def _typecheck(self, annotation):
```

Comment:
might also be nice if this were renamed appropriately so it's harder to think that `_typecheck` checks the entire value